libs/mve/bundle_io.cc

```diff
@@ -480,4 +483,537 @@ save_photosynther_bundle (Bundle::ConstPtr bundle, std::string const& filename)
+                continue;
+            view_ids.push_back(view_id);
+            Bundle::Feature2D corresponding_feature = view_id_to_features_2d.at(view_id).at(feature_2d_idx);
+            assert(corresponding_feature.feature_id == feature_3d_id);
+            Bundle::Feature2D ref;
+            ref.view_id = view_id;
+            ref.feature_id = feature_2d_idx;
+            ref.pos[0] = corresponding_feature.pos[0];
+            ref.pos[1] = corresponding_feature.pos[1];
+            refs.push_back(ref);
+            ++num_refs;
+        }
+
+        /* There should be at least 2 cameras that see the point. */
+        if (num_refs < 2 || num_refs > num_views)
+        {
+            throw util::Exception("Invalid number of feature refs: ",
+                util::string::get(num_refs));
+        }
+        feature_3d.refs = refs;
+        features.push_back(feature_3d);
+        ++num_points_3d;
+    }
+    in_points3D.close();
+}
+
+void
+read_colmap_cameras_bin_params(std::vector<float>& params,
+    std::string const& model, std::ifstream& in_cameras)
+{
+    if (model == "SIMPLE_PINHOLE")
+        params.resize(3);
+    else if (model == "PINHOLE" || model == "SIMPLE_RADIAL")
+        params.resize(4);
+    else
+        throw util::Exception("Unsupported camera model provided");
+    for (std::size_t param_idx = 0; param_idx < params.size(); ++param_idx)
+        params[param_idx] = (float)util::system::read_binary_little_endian<double>(&in_cameras);
+}
+
+void
+load_colmap_cameras_bin(std::string const& cameras_filename,
+    std::map<uint32_t, CameraInfo>& camera_colmap_id_to_info)
+{
+    std::cout << "Colmap: Loading cameras.bin file..." << std::endl;
+    define_camera_models();
+    std::ifstream in_cameras(cameras_filename.c_str());
+
+    uint32_t camera_colmap_id;  // starts at 1
+    std::string model_name;
+    uint32_t width, height;
+    std::vector<float> params;
+    uint64_t num_cam_models = util::system::read_binary_little_endian<uint64_t>(&in_cameras);
+    for (uint64_t cam_model_idx = 0; cam_model_idx < num_cam_models; ++cam_model_idx)
+    {
+        camera_colmap_id = util::system::read_binary_little_endian<camera_t>(&in_cameras);
+        camera_t model_code = util::system::read_binary_little_endian<int>(&in_cameras);
+        model_name = camera_model_code_to_name.at(model_code);
+        width = (uint32_t)util::system::read_binary_little_endian<uint64_t>(&in_cameras);
+        height = (uint32_t)util::system::read_binary_little_endian<uint64_t>(&in_cameras);
+        read_colmap_cameras_bin_params(params, model_name, in_cameras);
+        CameraInfo camera_info;
+        create_camera_info_from_params(
+            camera_info, model_name, params, width, height);
+        camera_colmap_id_to_info[camera_colmap_id] = camera_info;
+    }
+    in_cameras.close();
+}
+
+void
+load_colmap_images_bin(std::string const& images_filename,
```

Comment:
I assume you have looked into this, but this can't be merged with the txt version to avoid code duplication?

Reply:
I think that the merged result would be inconvenient to write and understand. 
However, in the latest commit of today I've wrapped most parts of the redundant code into specific functions that are now re-used to avoid code duplication. What do you think?